--- a/hw/kdrive/Makefile.am	2017-12-20 21:32:33.000000000 +0100
+++ b/hw/kdrive/Makefile.am	2020-01-20 17:51:45.597430805 +0100
@@ -6,6 +6,10 @@ if XFAKESERVER
 XFAKE_SUBDIRS = fake
 endif
 
+if X2GOKDRIVE
+X2GOKDRIVE_SUBDIRS = x2gokdrive
+endif
+
 if XEPHYR
 XEPHYR_SUBDIRS = ephyr
 endif
@@ -17,6 +21,7 @@ endif
 SERVER_SUBDIRS = 		\
 	$(FBDEV_SUBDIRS)	\
 	$(XEPHYR_SUBDIRS)       \
+	$(X2GOKDRIVE_SUBDIRS)	\
 	$(XFAKE_SUBDIRS)
 
 SUBDIRS =			\
@@ -24,7 +29,7 @@ SUBDIRS =			\
 	$(LINUX_SUBDIRS)	\
 	$(SERVER_SUBDIRS)
 
-DIST_SUBDIRS = fbdev ephyr src linux fake
+DIST_SUBDIRS = fbdev ephyr x2gokdrive src linux fake
 
 relink:
 	$(AM_V_at)for i in $(SERVER_SUBDIRS) ; do $(MAKE) -C $$i relink || exit 1 ; done
