--- a/hw/kdrive/Makefile.am	2018-10-25 16:13:21.000000000 +0200
+++ b/hw/kdrive/Makefile.am	2019-06-11 14:24:53.626300847 +0200
@@ -1,15 +1,20 @@
+if X2GOKDRIVE
+X2GOKDRIVE_SUBDIRS = x2gokdrive
+endif
+
 if XEPHYR
 XEPHYR_SUBDIRS = ephyr
 endif
 
 SERVER_SUBDIRS = 		\
-	$(XEPHYR_SUBDIRS)
+	$(XEPHYR_SUBDIRS)	\
+	$(X2GOKDRIVE_SUBDIRS)
 
 SUBDIRS =			\
 	src			\
 	$(SERVER_SUBDIRS)
 
-DIST_SUBDIRS = ephyr src
+DIST_SUBDIRS = ephyr x2gokdrive src
 
 relink:
 	$(AM_V_at)for i in $(SERVER_SUBDIRS) ; do $(MAKE) -C $$i relink || exit 1 ; done
